File: app/src/main/java/com/mert/bilgram/FeedActivity.java
package com.mert.bilgram;

import android.content.Context;
import android.content.DialogInterface;
import android.content.Intent;
import android.content.SharedPreferences;
import android.content.res.ColorStateList;
import android.database.Cursor;
import android.database.sqlite.SQLiteDatabase;
import android.database.sqlite.SQLiteStatement;
import android.graphics.Color;
import android.net.ConnectivityManager;
import android.net.NetworkInfo;
import android.os.Bundle;
import android.preference.PreferenceManager;
import android.view.Menu;
import android.view.MenuInflater;
import android.view.MenuItem;
import android.view.View;
import android.widget.EditText;
import android.widget.Toast;

import androidx.annotation.NonNull;
import androidx.appcompat.app.AlertDialog;
import androidx.appcompat.app.AppCompatActivity;
import androidx.appcompat.app.AppCompatDelegate;
import androidx.recyclerview.widget.DividerItemDecoration;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import com.google.android.material.floatingactionbutton.FloatingActionButton;
import com.google.firebase.auth.FirebaseAuth;
import com.google.firebase.database.DataSnapshot;
import com.google.firebase.database.DatabaseError;
import com.google.firebase.database.DatabaseReference;
import com.google.firebase.database.FirebaseDatabase;
import com.google.firebase.database.Query;
import com.google.firebase.database.ValueEventListener;
import com.google.firebase.storage.FirebaseStorage;
import com.google.firebase.storage.StorageReference;
import com.novoda.merlin.Connectable;
import com.novoda.merlin.Disconnectable;
import com.novoda.merlin.Merlin;

import java.text.SimpleDateFormat;
import java.util.ArrayList;
import java.util.Date;
import java.util.HashMap;
import java.util.Locale;

public class FeedActivity extends AppCompatActivity {

    FirebaseDatabase firebaseDatabase = FirebaseDatabase.getInstance();
    FirebaseStorage firebaseStorage = FirebaseStorage.getInstance();
    DatabaseReference myRef = firebaseDatabase.getReference();
    ArrayList<String> userEmailFromFB = new ArrayList<>(), userImageFromFB = new ArrayList<>(), userPostDescFromFB = new ArrayList<>(), postIDsFromFB = new ArrayList<>();
    RecyclerViewAdapter adapter = new RecyclerViewAdapter(userEmailFromFB, userPostDescFromFB, userImageFromFB, this, this);
    SharedPreferences mPrefs;
    FloatingActionButton fab;
    Merlin merlin;

    static SQLiteDatabase database;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

        mPrefs = PreferenceManager.getDefaultSharedPreferences(this);
        if (mPrefs.getBoolean("NIGHT_MODE", false))
            AppCompatDelegate.setDefaultNightMode(AppCompatDelegate.MODE_NIGHT_YES);
        else AppCompatDelegate.setDefaultNightMode(AppCompatDelegate.MODE_NIGHT_NO);

        if (AppCompatDelegate.getDefaultNightMode() == AppCompatDelegate.MODE_NIGHT_YES)
            setTheme(R.style.AppThemeDark);

        merlin = new Merlin.Builder().withAllCallbacks().build(this);

        merlin.registerConnectable(new Connectable() {
            @Override
            public void onConnect() {
                fab.setBackgroundTintList(ColorStateList.valueOf(FeedActivity.this.getColor(android.R.color.holo_red_dark)));
                fab.setImageDrawable(FeedActivity.this.getDrawable(R.drawable.ic_add_white_24dp));

                fab.setOnClickListener(new View.OnClickListener() {
                    @Override
                    public void onClick(View v) {
                        goToUploadActivity(v);
                    }
                });
            }
        });

        merlin.registerDisconnectable(new Disconnectable() {
            @Override
            public void onDisconnect() {
                fab.setBackgroundTintList(ColorStateList.valueOf(Color.DKGRAY));
                fab.setImageDrawable(getResources().getDrawable(android.R.drawable.stat_sys_warning));

                fab.setOnClickListener(new View.OnClickListener() {
                    @Override
                    public void onClick(View v) {
                        Toast.makeText(FeedActivity.this, FeedActivity.this.getString(R.string.connection_error), Toast.LENGTH_SHORT).show();
                    }
                });
            }
        });

        setContentView(R.layout.activity_feed);
        fab = findViewById(R.id.fab);

        ConnectivityManager cm = (ConnectivityManager) getSystemService(Context.CONNECTIVITY_SERVICE);

        NetworkInfo activeNetwork = cm.getActiveNetworkInfo();
        boolean isConnected = activeNetwork != null && activeNetwork.isConnected();

        if (isConnected) {

            getDataFromFirebase();

        } else {

            fab.setBackgroundTintList(ColorStateList.valueOf(Color.DKGRAY));
            fab.setImageDrawable(getResources().getDrawable(android.R.drawable.stat_sys_warning));

            fab.setOnClickListener(new View.OnClickListener() {
                @Override
                public void onClick(View v) {
                    Toast.makeText(FeedActivity.this, FeedActivity.this.getString(R.string.connection_error), Toast.LENGTH_SHORT).show();
                }
            });

            database = openOrCreateDatabase("Posts", MODE_PRIVATE, null);
            database.execSQL("CREATE TABLE IF NOT EXISTS posts (id VARCHAR, url VARCHAR, postDesc VARCHAR, useremail VARCHAR)");

            Cursor cursor = FeedActivity.database.rawQuery("SELECT * FROM posts ORDER BY id DESC", null);

            int IdIndex = cursor.getColumnIndex("id");
            int urlIndex = cursor.getColumnIndex("url");
            int postDescIndex = cursor.getColumnIndex("postDesc");
            int userEmailIndex = cursor.getColumnIndex("useremail");

            cursor.moveToFirst();

            do {

                postIDsFromFB.add(cursor.getString(IdIndex));
                userImageFromFB.add(cursor.getString(urlIndex));
                userEmailFromFB.add(cursor.getString(userEmailIndex));
                userPostDescFromFB.add(cursor.getString(postDescIndex));
                adapter.notifyDataSetChanged();

            } while (cursor.moveToNext());

            cursor.close();

        }

        RecyclerView recyclerView = findViewById(R.id.recyclerView);
        recyclerView.addItemDecoration(new DividerItemDecoration(this, DividerItemDecoration.VERTICAL));
        recyclerView.setAdapter(adapter);
        recyclerView.setLayoutManager(new LinearLayoutManager(this));

        ItemClickSupport.addTo(recyclerView).setOnItemClickListener(new ItemClickSupport.OnItemClickListener() {
            @Override
            public void onItemClicked(RecyclerView recyclerView, int position, View v) {
                Intent intent = new Intent(getApplicationContext(), ViewImageActivity.class);
                intent.putExtra("downloadURL", userImageFromFB.get(position));
                startActivity(intent);
            }

            @Override
            public void onItemDoubleClicked(RecyclerView recyclerView, int position, View v) {
                Toast.makeText(FeedActivity.this, "Liked", Toast.LENGTH_SHORT).show();

                String id = postIDsFromFB.get(position);
                String url = userImageFromFB.get(position);
                String postdesc = userPostDescFromFB.get(position);
                String useremail = userEmailFromFB.get(position);
                database = openOrCreateDatabase("SavedPosts", MODE_PRIVATE, null);
                database.execSQL("CREATE TABLE IF NOT EXISTS posts (id VARCHAR, url VARCHAR, postDesc VARCHAR, useremail VARCHAR, time VARCHAR)");

                String sqlString = "INSERT INTO posts (id, url, postDesc, useremail, time) VALUES (?, ?, ?, ?, ?)";
                SQLiteStatement statement = database.compileStatement(sqlString);
                statement.bindString(1, id);
                statement.bindString(2, url);
                statement.bindString(3, postdesc);
                statement.bindString(4, useremail);
                statement.bindString(5, new SimpleDateFormat("yyyyMMdd_HHmmss", Locale.US).format(new Date()));
                statement.execute();
            }
        });

    }

    public boolean itemLongClicked(final int i) {

        FirebaseAuth mAuth = FirebaseAuth.getInstance();

        if (mAuth.getCurrentUser().getEmail().equals(userEmailFromFB.get(i))) {

            final EditText edittext = new EditText(this);
            edittext.setText(userPostDescFromFB.get(i));

            AlertDialog.Builder builder = new AlertDialog.Builder(this);
            builder.setTitle("Edit or Delete")
                    .setView(edittext)
                    .setMessage("Do you want to edit or delete this post?")
                    .setPositiveButton("Delete", new DialogInterface.OnClickListener() {
                        @Override
                        public void onClick(DialogInterface dialogInterface, int i2) {

                            AlertDialog.Builder builder1 = new AlertDialog.Builder(FeedActivity.this);
                            builder1.setTitle("Are you sure?")
                                    .setMessage("You are about to delete this post permanently.")
                                    .setPositiveButton("Yes", new DialogInterface.OnClickListener() {
                                        @Override
                                        public void onClick(DialogInterface dialog, int which) {
                                            Query query = myRef.child("Posts").orderByChild("downloadURL").equalTo(userImageFromFB.get(i));
                                            query.addListenerForSingleValueEvent(new ValueEventListener() {
                                                @Override
                                                public void onDataChange(@NonNull DataSnapshot dataSnapshot) {
                                                    for (DataSnapshot snapshot : dataSnapshot.getChildren()) {
                                                        snapshot.getRef().removeValue();
                                                        StorageReference photoRef = firebaseStorage.getReferenceFromUrl(userImageFromFB.get(i));
                                                        photoRef.delete();
                                                    }

                                                    userImageFromFB.remove(i);
                                                    userPostDescFromFB.remove(i);
                                                    userEmailFromFB.remove(i);
                                                    adapter.notifyDataSetChanged();
                                                }

                                                @Override
                                                public void onCancelled(@NonNull DatabaseError databaseError) {
                                                }
                                            });

                                        }


                                    }).setNegativeButton("No", new DialogInterface.OnClickListener() {
                                @Override
                                public void onClick(DialogInterface dialog, int which) {

                                }
                            }).show();

                        }

                    }).setNeutralButton("EDIT", new DialogInterface.OnClickListener() {
                @Override
                public void onClick(DialogInterface dialogInterface, int i2) {

                    final String newDesc = edittext.getText().toString();

                    Query query = myRef.child("Posts").orderByChild("downloadURL").equalTo(userImageFromFB.get(i));

                    query.addListenerForSingleValueEvent(new ValueEventListener() {
                        @Override
                        public void onDataChange(@NonNull DataSnapshot dataSnapshot) {
                            for (DataSnapshot snapshot : dataSnapshot.getChildren()) {

                                DatabaseReference numMesasReference = snapshot.getRef().child("postDesc");
                                numMesasReference.setValue(newDesc);
                            }
                            userPostDescFromFB.set(i, newDesc);
                            adapter.notifyDataSetChanged();
                        }

                        @Override
                        public void onCancelled(@NonNull DatabaseError databaseError) {
                        }
                    });

                }
            }).show();

        }

        return true;

    }

    @Override
    public void onBackPressed() {
        moveTaskToBack(true);
    }

    public void getDataFromFirebase() {

        database = openOrCreateDatabase("Posts", MODE_PRIVATE, null);
        database.execSQL("CREATE TABLE IF NOT EXISTS posts (id VARCHAR, url VARCHAR, postDesc VARCHAR, useremail VARCHAR)");
        database.execSQL("DELETE FROM Posts");

        DatabaseReference newReference = firebaseDatabase.getReference("Posts");
        newReference.addValueEventListener(new ValueEventListener() {
            @Override
            public void onDataChange(@NonNull DataSnapshot dataSnapshot) {

                for (DataSnapshot ds : dataSnapshot.getChildren()) {

                    HashMap<String, String> hashMap = (HashMap<String, String>) ds.getValue();

                    if (userImageFromFB.contains(hashMap.get("downloadURL")) || hashMap.get("downloadURL") == null)
                        continue;

                    userEmailFromFB.add(0, hashMap.get("useremail"));
                    userPostDescFromFB.add(0, hashMap.get("postDesc"));
                    userImageFromFB.add(0, hashMap.get("downloadURL"));
                    postIDsFromFB.add(0, hashMap.get("ID"));

                    String sqlString = "INSERT INTO posts (id, url, postDesc, useremail) VALUES (?, ?, ?, ?)";
                    SQLiteStatement statement = database.compileStatement(sqlString);
                    statement.bindString(1, postIDsFromFB.get(0));
                    statement.bindString(2, userImageFromFB.get(0));
                    statement.bindString(3, userPostDescFromFB.get(0));
                    statement.bindString(4, userEmailFromFB.get(0));
                    statement.execute();

                    adapter.notifyDataSetChanged();
                }

            }

            @Override
            public void onCancelled(@NonNull DatabaseError databaseError) {

            }
        });

    }

    @Override
    public boolean onCreateOptionsMenu(Menu menu) {

        MenuInflater menuInflater = getMenuInflater();
        menuInflater.inflate(R.menu.feed_menu, menu);

        if (AppCompatDelegate.getDefaultNightMode() == AppCompatDelegate.MODE_NIGHT_YES)
            menu.getItem(3).setChecked(true);

        return super.onCreateOptionsMenu(menu);
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {

        switch (item.getItemId()) {

            case R.id.sign_out:
                FirebaseAuth user = FirebaseAuth.getInstance();
                user.signOut();
                startActivity(new Intent(getApplicationContext(), SignInWithTabs.class));
                break;

            case R.id.darkLight:

                item.setChecked(!item.isChecked());

                SharedPreferences.Editor editor = mPrefs.edit();
                editor.putBoolean("NIGHT_MODE", item.isChecked());
                editor.apply();

                startActivity(new Intent(this, FeedActivity.class));

                break;

            case R.id.myProfile:

                startActivity(new Intent(this, FavoritesActivity.class));
                break;

            case R.id.privacy_policy:

                startActivity(new Intent(this, PrivacyPolicyActivity.class));
                break;
        }

        return super.onOptionsItemSelected(item);
    }

    public void goToUploadActivity(View view) {

        Intent intent = new Intent(getApplicationContext(), UploadActivity.class);
        startActivity(intent);

    }

    @Override
    protected void onResume() {
        super.onResume();
        merlin.bind();
    }

}

File: app/src/main/java/com/mert/bilgram/FeedActivity.kt
package com.mert.bilgram

import android.annotation.SuppressLint
import android.content.Context
import android.content.Intent
import android.content.res.ColorStateList
import android.content.res.Configuration
import android.database.sqlite.SQLiteDatabase
import android.graphics.Color
import android.net.ConnectivityManager
import android.os.Bundle
import android.view.Menu
import android.view.MenuItem
import android.view.View
import android.widget.EditText
import androidx.appcompat.app.AlertDialog
import androidx.appcompat.app.AppCompatActivity
import androidx.core.content.ContextCompat
import androidx.recyclerview.widget.DividerItemDecoration
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import com.google.android.material.floatingactionbutton.FloatingActionButton
import com.google.android.material.snackbar.Snackbar
import com.google.firebase.auth.FirebaseAuth
import com.google.firebase.database.DataSnapshot
import com.google.firebase.database.DatabaseError
import com.google.firebase.database.FirebaseDatabase
import com.google.firebase.database.ValueEventListener
import com.google.firebase.storage.FirebaseStorage
import com.novoda.merlin.Merlin
import java.text.SimpleDateFormat
import java.util.*

class FeedActivity : AppCompatActivity() {
    private var firebaseDatabase = FirebaseDatabase.getInstance()
    var firebaseStorage = FirebaseStorage.getInstance()
    private var myRef = firebaseDatabase.reference
    var userEmailFromFB = ArrayList<String?>()
    var userImageFromFB = ArrayList<String?>()
    var userPostDescFromFB = ArrayList<String?>()
    var postIDsFromFB = ArrayList<String?>()
    var adapter = RecyclerViewAdapter(userEmailFromFB, userPostDescFromFB, userImageFromFB, this, this)
    private lateinit var fab: FloatingActionButton
    private lateinit var merlin: Merlin

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        when (resources.configuration.uiMode and Configuration.UI_MODE_NIGHT_MASK) {
            Configuration.UI_MODE_NIGHT_YES -> setTheme(R.style.AppThemeDark)
            Configuration.UI_MODE_NIGHT_NO -> setTheme(R.style.AppThemeLight)
        }

        merlin = Merlin.Builder().withAllCallbacks().build(this)

        merlin.registerConnectable {
            fab.backgroundTintList = ColorStateList.valueOf(getColor(android.R.color.holo_red_dark))
            fab.setImageDrawable(getDrawable(R.drawable.ic_add_white_24dp))
            fab.setOnClickListener { v -> goToUploadActivity(v) }
        }

        merlin.registerDisconnectable {
            fab.backgroundTintList = ColorStateList.valueOf(Color.DKGRAY)
            fab.setImageDrawable(ContextCompat.getDrawable(this, android.R.drawable.stat_sys_warning))
            fab.setOnClickListener { Snackbar.make(fab, R.string.connection_error, Snackbar.LENGTH_SHORT).show() }
        }

        setContentView(R.layout.activity_feed)

        fab = findViewById(R.id.fab)

        if (isConnectedToInternet())
            dataFromFirebase
        else {
            fab.backgroundTintList = ColorStateList.valueOf(Color.DKGRAY)

            fab.setImageDrawable(ContextCompat.getDrawable(this, android.R.drawable.stat_sys_warning))

            fab.setOnClickListener { Snackbar.make(fab, R.string.connection_error, Snackbar.LENGTH_SHORT).show() }

            database = openOrCreateDatabase("Posts", Context.MODE_PRIVATE, null)

            database.execSQL("CREATE TABLE IF NOT EXISTS posts (id VARCHAR, url VARCHAR, postDesc VARCHAR, useremail VARCHAR)")

            val cursor = database.rawQuery("SELECT * FROM posts ORDER BY id DESC", null)

            val idIndex = cursor.getColumnIndex("id")

            val urlIndex = cursor.getColumnIndex("url")

            val postDescIndex = cursor.getColumnIndex("postDesc")

            val userEmailIndex = cursor.getColumnIndex("useremail")

            cursor.moveToFirst()

            do {
                postIDsFromFB.add(cursor.getString(idIndex))
                userImageFromFB.add(cursor.getString(urlIndex))
                userEmailFromFB.add(cursor.getString(userEmailIndex))
                userPostDescFromFB.add(cursor.getString(postDescIndex))
                adapter.notifyDataSetChanged()

            } while (cursor.moveToNext())

            cursor.close()
        }

        val recyclerView = findViewById<RecyclerView>(R.id.recyclerView)

        recyclerView.addItemDecoration(DividerItemDecoration(this, DividerItemDecoration.VERTICAL))

        recyclerView.adapter = adapter

        recyclerView.layoutManager = LinearLayoutManager(this)

        ItemClickSupport.addTo(recyclerView).setOnItemClickListener(object : ItemClickSupport.OnItemClickListener {
            override fun onItemClicked(recyclerView: RecyclerView?, position: Int, v: View?) {
                startActivity(Intent(applicationContext, ViewImageActivity::class.java).putExtra("downloadURL", userImageFromFB[position]))
            }

            override fun onItemDoubleClicked(recyclerView: RecyclerView?, position: Int, v: View?) {
                Snackbar.make(v!!, R.string.liked, Snackbar.LENGTH_SHORT).show()

                database = openOrCreateDatabase("SavedPosts", Context.MODE_PRIVATE, null)
                database.execSQL("CREATE TABLE IF NOT EXISTS posts (id VARCHAR, url VARCHAR, postDesc VARCHAR, useremail VARCHAR, time VARCHAR)")

                val statement = database.compileStatement("INSERT INTO posts (id, url, postDesc, useremail, time) VALUES (?, ?, ?, ?, ?)")

                statement.bindString(1, postIDsFromFB[position])
                statement.bindString(2, userImageFromFB[position])
                statement.bindString(3, userPostDescFromFB[position])
                statement.bindString(4, userEmailFromFB[position])
                statement.bindString(5, SimpleDateFormat("yyyyMMdd_HHmmss", Locale.US).format(Date()))
                statement.execute()
            }
        })
    }

    @SuppressLint("MissingPermission")
    private fun isConnectedToInternet(): Boolean {
        (getSystemService(Context.CONNECTIVITY_SERVICE) as ConnectivityManager).run {
            this.getNetworkCapabilities(this.activeNetwork)?.run {
                return true
            }
        }

        return false
    }

    fun itemLongClicked(i: Int): Boolean {
        val mAuth = FirebaseAuth.getInstance()
        if (mAuth.currentUser!!.email == userEmailFromFB[i]) {

            val editText = EditText(this)
            editText.setText(userPostDescFromFB[i])

            AlertDialog.Builder(this).setTitle(R.string.edit_or_delete)
                    .setView(editText)
                    .setMessage(R.string.edit_or_delete_desc)
                    .setPositiveButton(R.string.delete) { _, _ ->
                        AlertDialog.Builder(this)
                                .setTitle(R.string.are_you_sure)
                                .setMessage(R.string.you_are_deleting_post_permanently)
                                .setPositiveButton(R.string.yes) { _, _ ->
                                    myRef.child("Posts").orderByChild("downloadURL").equalTo(userImageFromFB[i])
                                            .addListenerForSingleValueEvent(object : ValueEventListener {
                                                override fun onDataChange(dataSnapshot: DataSnapshot) {
                                                    for (snapshot in dataSnapshot.children) {
                                                        snapshot.ref.removeValue()
                                                        firebaseStorage.getReferenceFromUrl(userImageFromFB[i]!!).delete()
                                                    }
                                                    userImageFromFB.removeAt(i)
                                                    userPostDescFromFB.removeAt(i)
                                                    userEmailFromFB.removeAt(i)
                                                    adapter.notifyItemRemoved(i)
                                                }

                                                override fun onCancelled(databaseError: DatabaseError) {}
                                            })
                                }.setNegativeButton(R.string.no) { _, _ -> }.show()

                    }.setNeutralButton(R.string.edit) { _, _ ->
                        val newDesc = editText.text.toString()
                        myRef.child("Posts").orderByChild("downloadURL").equalTo(userImageFromFB[i])
                                .addListenerForSingleValueEvent(object : ValueEventListener {
                                    override fun onDataChange(dataSnapshot: DataSnapshot) {
                                        for (snapshot in dataSnapshot.children)
                                            snapshot.ref.child("postDesc").setValue(newDesc)

                                        userPostDescFromFB[i] = newDesc
                                        adapter.notifyItemChanged(i)
                                    }

                                    override fun onCancelled(databaseError: DatabaseError) {}
                                })
                    }.show()
        }
        return true
    }

    override fun onBackPressed() {
        moveTaskToBack(true)
    }

    private val dataFromFirebase: Unit
        get() {
            database = openOrCreateDatabase("Posts", Context.MODE_PRIVATE, null)
            database.execSQL("CREATE TABLE IF NOT EXISTS posts (id VARCHAR, url VARCHAR, postDesc VARCHAR, useremail VARCHAR)")
            database.execSQL("DELETE FROM Posts")

            firebaseDatabase.getReference("Posts").addValueEventListener(object : ValueEventListener {
                override fun onDataChange(dataSnapshot: DataSnapshot) {
                    for (ds in dataSnapshot.children) {
                        val hashMap = ds.value as HashMap<*, *>
                        if (userImageFromFB.contains(hashMap["downloadURL"]) || hashMap["downloadURL"] == null)
                            continue

                        userEmailFromFB.add(0, hashMap["useremail"] as String?)
                        userPostDescFromFB.add(0, hashMap["postDesc"] as String?)
                        userImageFromFB.add(0, hashMap["downloadURL"] as String?)
                        postIDsFromFB.add(0, hashMap["ID"] as String?)

                        val statement = database.compileStatement("INSERT INTO posts (id, url, postDesc, useremail) VALUES (?, ?, ?, ?)")

                        statement.bindString(1, postIDsFromFB[0])
                        statement.bindString(2, userImageFromFB[0])
                        statement.bindString(3, userPostDescFromFB[0])
                        statement.bindString(4, userEmailFromFB[0])
                        statement.execute()
                        adapter.notifyDataSetChanged()
                    }
                }

                override fun onCancelled(databaseError: DatabaseError) {}
            })
        }

    override fun onCreateOptionsMenu(menu: Menu): Boolean {
        menuInflater.inflate(R.menu.feed_menu, menu)
        return super.onCreateOptionsMenu(menu)
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        when (item.itemId) {
            R.id.sign_out -> {
                FirebaseAuth.getInstance().signOut()
                startActivity(Intent(this, SignInWithTabs::class.java))
            }
            R.id.myProfile -> startActivity(Intent(this, FavoritesActivity::class.java))
            R.id.privacy_policy -> startActivity(Intent(this, PrivacyPolicyActivity::class.java))
        }
        return super.onOptionsItemSelected(item)
    }

    fun goToUploadActivity(view: View?) {
        startActivity(Intent(applicationContext, UploadActivity::class.java))
    }

    override fun onResume() {
        super.onResume()
        merlin.bind()
    }

    companion object {
        lateinit var database: SQLiteDatabase
    }
}